### ISSUE TITLE
Support generic typedef

### ISSUE THREAD

**Issue by andy-ms** [Microsoft/TypeScript#20831]
**TypeScript Version:**  2.7.0-dev.20171220

**Code**

```ts
/** @typedef {T | null | undefined} Maybe<T> */

/** @type Maybe<number> */
const x = 0;
```

**Expected behavior:**

No error.

**Actual behavior:**

```
a.js(1,15): error TS2304: Cannot find name 'T'.
a.js(3,11): error TS2315: Type 'Maybe' is not generic.
```

**Comment by weswigham** [comment]
@andy-ms In your example, I'm pretty sure you need to introduce the `T` in the `@typedef` with an `@template` tag, too.

**Comment by brendankenny** [comment]
I believe this is now fixed (by #24600?). @weswigham's closure example
```js
/**
 * @template T
 * @typedef {T | null | undefined} Maybe
 */

/** @type Maybe<number> */
const x = 0;
```
now works in latest.

**Comment by mhegazy** [comment]
@brendankenny's comment in https://github.com/Microsoft/TypeScript/issues/20831#issuecomment-400427305 is the expected format to declare a generic type parameter in a type def.
